### ISSUE TITLE
[question] Why Do I Need to Directly List a Header-Only Library Dependency When Using a Static/Shared Library?

### ISSUE THREAD

**Issue by astough** [conan-io/conan#16965]
### What is your question?

Hi. I have a header-only library that is being packaged with Conan. This header-only library is a dependency of another Conan package that builds a static/shared library based upon the `options = "shared": [True, False]` attribute. This static/shared library is a Conan dependency of a application that builds an executable.

To properly compile my executable, I have to list both the header-only library and static/shared library in my Conanfile (e.g., `requires = "my_header_only_library/1.0.0", "my_static_shared_library/1.0.0")`. If I only include the dependency for my static/shared library, my application is unable to compile because it can't include the header files in the header-only library. This seems to be related to the `options = "shared": [True, False]` attribute, because if I remove that attribute from my static/shared library (the package is now just a static library), I'm able to just specify the dependency to my library in my Conanfile and compile my application.

What is the `options = "shared": [True, False]` attribute changing that causes the dependency of my dependency (i.e., a header file) to not be included by my application? Is there some other way to include the header file contained within the header-only library in my application so I don't need to specify the dependency twice (i.e., once in the static/shared library, once in the application)?

Thanks!

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @astough 

Thanks for your question.

For transitive dependency headers there are 3 different possibilities:

- if the consumer also contains explicit ``#include`` to the transitive dependency, then it is not really a transitive dependency, and it must be explicitly required by the consumer with a ``requires()``
- If the consumer does not contain explicit ``#includes`` to the transitive dependency, but it happens that the transitive shared dependency has ``#includes`` to the transitive library in their **public** package headers, then it should declare it explicitly as ``requires("header_only_pkg/version", transitive_headers=True)`` because the default is assuming dependencies headers are private implementation details.
- finally if the consumer does not contain explicit ``#includes`` to the transitive dependency, and the transitive shared dependency does not have ``#includes`` to the transitive library in their **public** package headers, then it is the Conan default, as the consumer really doesn't need visibility over those headers.

So basically adding ``transitive_headers=True`` in the requirement from the shared lib to the header lib should fix it.
Please try it and let us know. Thanks!

**Comment by astough** [comment]
I've added `requires("my_header_only_library/1.0.0", transitive_headers=True)` to the Conanfile for my static/shared library and now my application is able to include the header files in my header-only library and compile. Previously I had tried to define `transitive_headers=True` for my header-only library in the header-only package itself using the `package_type = "header-library"` attribute but that didn't seem to work. I didn't think about specifying `transitive_headers=True` when pulling in the header-only library to my static/shared library.

**Comment by memsharded** [comment]
> Previously I had tried to define transitive_headers=True for my header-only library in the header-only package itself using the package_type = "header-library" attribute but that didn't seem to work. 

Yes, those are different concepts. One thing is the package type, being a header-only library, but how those headers are propagated down can only be known by the user of this header-library package, because it depends if the headers are used only from private .h or .cpp files or if the headers are used by public headers of the consumer package. So the ``requires(..., transitive_header=True)`` can only be determined by the consumer.

**Comment by astough** [comment]
Thanks for the clarification.

**Comment by memsharded** [comment]
Happy to help, don't hesitate to create new tickets for any further question. Thanks for the feedback.
